# **Applied Machine Learning Homework 5**
**Due 2 May, 2022 (Monday) 11:59PM EST**

**Name: Aayush Kumar Verma** <br>
**UNI: av2955**

### Natural Language Processing
We will train a supervised training model to predict if a tweet has a positive or negative sentiment.

####  **Dataset loading & dev/test splits**

**1.1) Load the twitter dataset from NLTK library**

In [127]:
import nltk
nltk.download('twitter_samples')
from nltk.corpus import twitter_samples 

[nltk_data] Downloading package twitter_samples to /root/nltk_data...
[nltk_data]   Package twitter_samples is already up-to-date!


**1.2) Load the positive & negative tweets**

In [128]:
all_positive_tweets = twitter_samples.strings('positive_tweets.json')
all_negative_tweets = twitter_samples.strings('negative_tweets.json')

**1.3) Create a development & test split (80/20 ratio):**

In [129]:
#code here

import re
import nltk
import string
from sklearn.model_selection import train_test_split
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression

In [130]:
x_dev_positive, x_test_positive, x_dev_negative, x_test_negative = train_test_split(all_positive_tweets, all_negative_tweets, random_state = 42, test_size=0.2)
x_dev = x_dev_positive + x_dev_negative
x_test = x_test_positive + x_test_negative
y_dev = [1] * len(x_dev_positive) + [0] * len(x_dev_negative)
y_test = [1] * len(x_test_positive) + [0] * len(x_test_negative)

In [131]:
print (x_dev[:10])
print ('--------')
print (x_test[:10])

['@seananmcguire my best friend Carina is one here in San Francisco. Let me know if you want to get I touch. :)', 'That also means, imma go back to being more twitter active :D\nCause I know everyone missed me ;) xD', '@MsKristinKreuk Hugs ang Kisses from the philippines :)', 'Please comment on what you think of what I wrote so far in the twitlonger :)', 'This girl :)  https://t.co/OAXMGGICNr', "@grrl_afraid Oh course, I gained weight bcus I used to be a rubbish veggie but now I've learned to balance everything now! :-)", '@Satbains1 Congratulations Sat on your 3rd honorary degree from everyone at Alacer :)', '@Geordnet hehe cool :)', '@Wilma2207fWilma haha! Thank you! :) 🍰', 'Hi BAM ! @BarsAndMelody \nCan you follow my bestfriend @969Horan696 ? \nShe loves you a lot :) \nSee you in Warsaw &lt;3 \nLove you &lt;3 x10']
--------
['@Sanza_T @LeratoTmohale Does the bag do the job Sanza?  :-)', '@WeAlwaysChillin This sounds nice :D', "@wishinguponstar We couldn't have said it better ourselv

#### **Data preprocessing**
We will do some data preprocessing before we tokenize the data. We will remove `#` symbol, hyperlinks, stop words & punctuations from the data. You can use the `re` package in python to find and replace these strings. 

**1.4) Replace the `#` symbol with '' in every tweet**

In [132]:
#code here

x_dev = [word.replace("#", "") for word in x_dev]
x_test = [word.replace("#", "") for word in x_test]

In [133]:
print (x_dev[:10])
print ('--------')
print (x_test[:10])

['@seananmcguire my best friend Carina is one here in San Francisco. Let me know if you want to get I touch. :)', 'That also means, imma go back to being more twitter active :D\nCause I know everyone missed me ;) xD', '@MsKristinKreuk Hugs ang Kisses from the philippines :)', 'Please comment on what you think of what I wrote so far in the twitlonger :)', 'This girl :)  https://t.co/OAXMGGICNr', "@grrl_afraid Oh course, I gained weight bcus I used to be a rubbish veggie but now I've learned to balance everything now! :-)", '@Satbains1 Congratulations Sat on your 3rd honorary degree from everyone at Alacer :)', '@Geordnet hehe cool :)', '@Wilma2207fWilma haha! Thank you! :) 🍰', 'Hi BAM ! @BarsAndMelody \nCan you follow my bestfriend @969Horan696 ? \nShe loves you a lot :) \nSee you in Warsaw &lt;3 \nLove you &lt;3 x10']
--------
['@Sanza_T @LeratoTmohale Does the bag do the job Sanza?  :-)', '@WeAlwaysChillin This sounds nice :D', "@wishinguponstar We couldn't have said it better ourselv

**1.5) Replace hyperlinks with '' in every tweet**

In [134]:
#code here

regex = 'https?://[a-z]*.[a-z]*(/[a-zA-Z0-9]*)?'
x_dev = [re.sub(regex, repl='', string=word) for word in x_dev]
x_test = [re.sub(regex, repl='', string=word) for word in x_test]

In [135]:
print (x_dev[:10])
print ('--------')
print (x_test[:10])

['@seananmcguire my best friend Carina is one here in San Francisco. Let me know if you want to get I touch. :)', 'That also means, imma go back to being more twitter active :D\nCause I know everyone missed me ;) xD', '@MsKristinKreuk Hugs ang Kisses from the philippines :)', 'Please comment on what you think of what I wrote so far in the twitlonger :)', 'This girl :)  ', "@grrl_afraid Oh course, I gained weight bcus I used to be a rubbish veggie but now I've learned to balance everything now! :-)", '@Satbains1 Congratulations Sat on your 3rd honorary degree from everyone at Alacer :)', '@Geordnet hehe cool :)', '@Wilma2207fWilma haha! Thank you! :) 🍰', 'Hi BAM ! @BarsAndMelody \nCan you follow my bestfriend @969Horan696 ? \nShe loves you a lot :) \nSee you in Warsaw &lt;3 \nLove you &lt;3 x10']
--------
['@Sanza_T @LeratoTmohale Does the bag do the job Sanza?  :-)', '@WeAlwaysChillin This sounds nice :D', "@wishinguponstar We couldn't have said it better ourselves, Kelly! So glad you'

**1.6) Remove all stop words**

In [136]:
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [137]:
#code here

stopword = nltk.corpus.stopwords.words('english')
x_dev = [[word for word in word_tokenize(tweet) if word not in stopword] for tweet in x_dev]
x_test = [[word for word in word_tokenize(tweet) if word not in stopword] for tweet in x_test]

**1.7) Remove all punctuations**

In [138]:
#code here

x_dev = [' '.join(tweet).translate(str.maketrans('','',string.punctuation)) for tweet in x_dev]
x_test = [' '.join(tweet).translate(str.maketrans('','',string.punctuation)) for tweet in x_test]

In [139]:
print (x_dev[:10])
print ("----------")
print (x_test[:10])

[' seananmcguire best friend Carina one San Francisco  Let know want get I touch   ', 'That also means  imma go back twitter active  D Cause I know everyone missed   xD', ' MsKristinKreuk Hugs ang Kisses philippines  ', 'Please comment think I wrote far twitlonger  ', 'This girl  ', ' grrlafraid Oh course  I gained weight bcus I used rubbish veggie I ve learned balance everything    ', ' Satbains1 Congratulations Sat 3rd honorary degree everyone Alacer  ', ' Geordnet hehe cool  ', ' Wilma2207fWilma haha  Thank    🍰', 'Hi BAM   BarsAndMelody Can follow bestfriend  969Horan696  She loves lot   See Warsaw  lt  3 Love  lt  3 x10']
----------
[' SanzaT  LeratoTmohale Does bag job Sanza    ', ' WeAlwaysChillin This sounds nice  D', ' wishinguponstar We could nt said better  Kelly  So glad re enjoying nibbles  ', 'Chaerin unnie  ', 'A productive day far   Happy Days        head spinning things  voices keep calling    Lol \U000fe334', ' twentyonepilots  fujirockjp tylers hipster glasses  D', '

**1.8) Apply stemming on the development & test datasets using Porter algorithm**

In [140]:
#code here
stemmer = nltk.stem.PorterStemmer()
x_dev = [' '.join([stemmer.stem(word) for word in word_tokenize(tweet)]) for tweet in x_dev]
x_test = [' '.join([stemmer.stem(word) for word in word_tokenize(tweet)]) for tweet in x_test]

In [141]:
print (x_dev[:10])
print ('--------')
print (x_test[:10])

['seananmcguir best friend carina one san francisco let know want get I touch', 'that also mean imma go back twitter activ D caus I know everyon miss xD', 'mskristinkreuk hug ang kiss philippin', 'pleas comment think I wrote far twitlong', 'thi girl', 'grrlafraid Oh cours I gain weight bcu I use rubbish veggi I ve learn balanc everyth', 'satbains1 congratul sat 3rd honorari degre everyon alac', 'geordnet hehe cool', 'wilma2207fwilma haha thank 🍰', 'Hi bam barsandmelodi can follow bestfriend 969horan696 she love lot see warsaw lt 3 love lt 3 x10']
--------
['sanzat leratotmohal doe bag job sanza', 'wealwayschillin thi sound nice D', 'wishinguponstar We could nt said better kelli So glad re enjoy nibbl', 'chaerin unni', 'A product day far happi day head spin thing voic keep call lol \U000fe334', 'twentyonepilot fujirockjp tyler hipster glass D', 'clivewalk near enough dartmoor', 'simonxxx follback', 'stat day arriv 2 new follow NO unfollow via', 'rofl shade grey netsec edit']


#### **Model training**

**1.9) Create bag of words features for each tweet in the development dataset**

In [142]:
#code here

vector_bow = CountVectorizer()
x_dev_bow = vector_bow.fit_transform(x_dev)
x_test_bow = vector_bow.transform(x_test)

**1.10) Train a supervised learning model of choice on the development dataset**

In [143]:
#code here

model = LogisticRegression()
model.fit(x_dev_bow, y_dev)

LogisticRegression()

**1.11) Create TF-IDF features for each tweet in the development dataset**

In [144]:
#code here

tfidf = TfidfVectorizer()
tfidf_dev = tfidf.fit_transform(x_dev)
tfidf_test = tfidf.transform(x_test)

**1.12) Train the same supervised learning algorithm on the development dataset with TF-IDF features**

In [145]:
#code here

model_tfidf = LogisticRegression()
model_tfidf.fit(tfidf_dev, y_dev)

LogisticRegression()

**1.13) Compare the performance of the two models on the test dataset**

In [146]:
print (x_test_bow)

  (0, 1333)	1
  (0, 3609)	1
  (0, 6618)	1
  (1, 9067)	1
  (1, 11889)	1
  (1, 12820)	1
  (2, 1648)	1
  (2, 2924)	1
  (2, 4046)	1
  (2, 5010)	1
  (2, 9063)	1
  (2, 9268)	1
  (2, 10546)	1
  (2, 11049)	1
  (2, 11770)	1
  (2, 13806)	1
  (3, 13413)	1
  (4, 2195)	1
  (4, 3261)	2
  (4, 4329)	1
  (4, 5392)	1
  (4, 5487)	1
  (4, 6951)	1
  (4, 7645)	1
  (4, 10273)	1
  :	:
  (1994, 10273)	1
  (1994, 10546)	1
  (1994, 13359)	1
  (1994, 13385)	1
  (1994, 13471)	1
  (1995, 7762)	2
  (1995, 8719)	1
  (1995, 11770)	1
  (1995, 14494)	1
  (1996, 8719)	1
  (1996, 9268)	1
  (1996, 12346)	1
  (1996, 13863)	1
  (1997, 11168)	1
  (1997, 11826)	1
  (1998, 4616)	1
  (1998, 4936)	1
  (1998, 9665)	1
  (1999, 2414)	1
  (1999, 4630)	1
  (1999, 7259)	1
  (1999, 7671)	1
  (1999, 9443)	1
  (1999, 12206)	1
  (1999, 13013)	1


In [147]:
#code here

model.score (x_test_bow, y_test)

0.7595

In [148]:
model_tfidf.score (tfidf_test, y_test)

0.766

##### TFIDF model performs better than the BOW model as seen from the score values of the models (0.766: TFIDF vs. 0.7595: BOW)